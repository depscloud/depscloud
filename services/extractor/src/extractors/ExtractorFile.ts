import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
