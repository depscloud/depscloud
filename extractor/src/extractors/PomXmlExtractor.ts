import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import Languages from "./Languages";
import MatchConfig from "../matcher/MatchConfig";
import {ManifestDependency, ManifestFile} from "@depscloud/api/v1beta";
import cheerio = require("cheerio");

export default class PomXmlExtractor implements Extractor {
    public matchConfig(): MatchConfig {
        return {
            includes: [
                "**/pom.xml",
                "**/*.pom",
            ],
            excludes: [
                "**/testdata/**",
            ],
        };
    }

    public requires(): string[] {
        return [ "pom.xml" ];
    }

    public async extract(_: string, files: { [p: string]: ExtractorFile }): Promise<ManifestFile> {
        const xml = files["pom.xml"].xml();

        const parentGroupId = xml.find("project > parent > groupId").text();
        const parentArtifactId = xml.find("project > parent > artifactId").text();
        const parentVersion = xml.find("project > parent > version").text();

        const groupId = xml.find("project > groupId").text() || parentGroupId;
        const artifactId = xml.find("project > artifactId").text();
        const version = xml.find("project > version").text();

        const sourceUrl = xml.find("project > scm > url").text();

        const dependencies: ManifestDependency[] = [];
        if (parentGroupId && parentArtifactId && parentVersion) {
            dependencies.push({
                name: [parentGroupId, parentArtifactId].join(":"),
                versionConstraint: parentVersion,
                scopes: [ "parent" ],
            });
        }

        const matched: Cheerio = xml.find("project > dependencies > dependency");
        matched.map((i, match: CheerioElement) => {
            const depGroupId = cheerio(match).find("> groupId").text();
            const depArtifactId = cheerio(match).find("> artifactId").text();
            const versionConstraint = cheerio(match).find("> version").text();
            const scope = cheerio(match).find("> scope").text();

            const scopes = [scope || "compile"];

            dependencies.push({
                name: [depGroupId, depArtifactId].join(":"),
                versionConstraint,
                scopes,
            });
        });

        return {
            language: Languages.JAVA,
            system: "maven",
            sourceUrl,
            name: [groupId, artifactId].join(":"),
            version,
            dependencies,
        };
    }
}
