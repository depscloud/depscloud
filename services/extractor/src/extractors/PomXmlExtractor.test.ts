import {readFile} from "fs";
import {promisify} from "util";
import ExtractorFile from "./ExtractorFile";
import PomXmlExtractor from "./PomXmlExtractor";

const readFileAsync = promisify(readFile);

describe("PomXmlExtractor", () => {
    test("fullParse", async () => {
        const pomPath = require.resolve("./testdata/pom.xml");
        const buffer = await readFileAsync(pomPath);
        const content = buffer.toString();

        const parser = new PomXmlExtractor();

        const actual = await parser.extract("", { "pom.xml": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
