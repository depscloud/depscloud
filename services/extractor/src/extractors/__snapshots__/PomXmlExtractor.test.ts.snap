// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PomXmlExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "parentGroupId:parentArtifactId",
      "scopes": Array [
        "parent",
      ],
      "versionConstraint": "1.0",
    },
    Object {
      "name": "compileGroupId:compileArtifactId",
      "scopes": Array [
        "compile",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "runtimeGroupId:runtimeArtifactId",
      "scopes": Array [
        "runtime",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "testGroupId:testArtifactId",
      "scopes": Array [
        "test",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "systemGroupId:systemArtifactId",
      "scopes": Array [
        "system",
      ],
      "versionConstraint": "1.0.0",
    },
    Object {
      "name": "providedGroupId:providedArtifactId",
      "scopes": Array [
        "provided",
      ],
      "versionConstraint": "1.0.0",
    },
  ],
  "language": "java",
  "name": "groupId:artifactId",
  "sourceUrl": "git@github.com:group/project.git",
  "system": "maven",
  "version": "9.9.9",
}
`;

exports[`PomXmlExtractor fullParse 2`] = `
"{
  \\"language\\": \\"java\\",
  \\"system\\": \\"maven\\",
  \\"sourceUrl\\": \\"git@github.com:group/project.git\\",
  \\"name\\": \\"groupId:artifactId\\",
  \\"version\\": \\"9.9.9\\",
  \\"dependencies\\": [
    {
      \\"name\\": \\"parentGroupId:parentArtifactId\\",
      \\"versionConstraint\\": \\"1.0\\",
      \\"scopes\\": [
        \\"parent\\"
      ]
    },
    {
      \\"name\\": \\"compileGroupId:compileArtifactId\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"compile\\"
      ]
    },
    {
      \\"name\\": \\"runtimeGroupId:runtimeArtifactId\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"runtime\\"
      ]
    },
    {
      \\"name\\": \\"testGroupId:testArtifactId\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"test\\"
      ]
    },
    {
      \\"name\\": \\"systemGroupId:systemArtifactId\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"system\\"
      ]
    },
    {
      \\"name\\": \\"providedGroupId:providedArtifactId\\",
      \\"versionConstraint\\": \\"1.0.0\\",
      \\"scopes\\": [
        \\"provided\\"
      ]
    }
  ]
}"
`;
